pkg/kafka/policy.go

```diff
@@ -281,9 +222,42 @@ func (req *RequestMessage) ruleMatches(rule api.PortRuleKafka) bool {
 // rules. The function will return true if the policy allows the message,
 // otherwise false is returned.
 func (req *RequestMessage) MatchesRule(rules []api.PortRuleKafka) bool {
+	reqTopicsMap := map[string]bool{}
+
+	// Maintain a map of all topics in the request
+	// Only if all the topics in a request are
+	// allowed by the rules, we should allow a request.
+	topics := req.GetTopics()
+	for _, topic := range topics {
+		reqTopicsMap[topic] = true
+	}
+
 	for _, rule := range rules {
-		if req.ruleMatches(rule) {
-			return true
+		switch len(topics) {
+		case 0:
+			if req.ruleMatches(rule, "") {
+				return true
+			}
+		default:
+			if rule.Topic != "" {
+				_, ok := reqTopicsMap[rule.Topic]
+				if ok && req.ruleMatches(rule, rule.Topic) {
+					delete(reqTopicsMap, rule.Topic)
+					// If all topics in a request are allowed by the
+					// set of rules we have traversed so far, along with the other
+					// fields in the request, the request
+					// should be allowed.
+					if len(reqTopicsMap) == 0 {
+						return true
+					}
+				}
+			} else {
+				// topic in rule is empty,
+				// check if other fields of the rule match.
+				if req.ruleMatches(rule, "") {
+					return true
+				}
+			}
 		}
```

Comment:
Overall, after removing all the `matchTopic` parameters and all the tests on `rule.Topic`, I think this whole method can be simplified as:
```
topics := req.GetTopics()
reqTopicsMap := make(map[string]bool, len(topics))
for _, topic := range topics {
    reqTopicsMap[topic] = true
}
for _, rule := range rules {
    if rule.Topic == "" {
        if req.ruleMatches(rule) {
            return true
        }
    } else if reqTopicsMap[rule.Topic] {
        if req.ruleMatches(rule) {
            delete(reqTopicsMap, rule.Topic)
            if len(reqTopicsMap) == 0 {
                return true
            }
        }
    }
}
return false
```

This also is more efficient as it's not calling `req.ruleMatches` if we know the rule's topic won't match or we've already matched the topic.

Reply:
Problem with this is it does not cover the cases with requests not containing a `topic` like 
`apiVersions` `findCoordinator` et al. That's why the special handling was in place. What do you suggest?


```
topics := req.GetTopics()
reqTopicsMap := make(map[string]bool, len(topics))
for _, topic := range topics {
    reqTopicsMap[topic] = true
}
	for _, rule := range rules {
		if rule.Topic == "" {
			if req.ruleMatches(rule) {
				return true
			}
		} else if reqTopicsMap[rule.Topic] {
			if req.ruleMatches(rule) {
				delete(reqTopicsMap, rule.Topic)
				if len(reqTopicsMap) == 0 {
					return true
				}
			}
		} else if len(topics) == 0 {
			if req.ruleMatches(rule) {
				return true
			}
		}
	}
return false
```

Reply:
That's not necessary. In case `len(topics) == 0`, we should only check rules where `rule.Topic == ""`, and that's already handled by the 1st `if`. The 2nd `if` condition will never be true in that case, and that's fine.

Reply:
done